virttest/utils_zchannels.py

```diff
@@ -0,0 +1,168 @@
+#
+# See LICENSE for more details.
+#
+# Copyright: Red Hat Inc. 2020
+# Author: Sebastian Mitterle <smitterl@redhat.com>
+
+"""
+Module for channel device management on s390x aka IBM Z
+
+Example:
+    # Retrieve device to remove it from host
+
+    subchannel_paths = SubchannelPaths()
+    subchannel_paths.get_info()
+
+    device = subchannel_paths.get_first_unused_and_safely_removable()
+    chpids = device[subchannel_paths.header["CHPIDs"]]
+
+    channel_paths = ChannelPaths()
+    channel_paths.set_standby(chipds)
+"""
+from virttest.utils_misc import cmd_status_output
+
+# timeout value in seconds for any command run in this module
+CMD_TIMEOUT = 60
+# for debugging, set True to get verbose command run output
+VERBOSE = False
+
+
+class SubchannelPaths(object):
+    """
+    Represents channel subsystem info
+    """
+
+    def __init__(self, session=None):
+        """
+        Initializes instance
+
+        :param session: guest session; if None, host info is handled
+        """
+        self.session = session
+        self.header = {
+                "Device": 0,
+                "Subchan.": 1,
+                "DevType": 2,
+                "CU Type": 3,
+                "Use": 4,
+                "PIM": 5,
+                "PAM": 6,
+                "POM": 7,
+                "CHPIDs": 8,
+                "CHPIDs_extra": 9
+                }
+        self.devices = []
+
+    def get_info(self):
+        """
+        Calls lscss and stores lines
+        """
+        err, out = cmd_status_output("lscss", shell=True,
+                                     session=self.session,
+                                     timeout=CMD_TIMEOUT,
+                                     verbose=VERBOSE)
+        if err:
+            raise OSError("Error when running lscss: %s" % out)
+        # skip header and split according to self.header
+        self.devices = [[
+            x[:8],
+            x[9:17],
+            x[19:26],
+            x[27:34],
+            x[35:38],
+            x[40:42],
+            x[44:46],
+            x[48:50],
+            x[53:61],
+            x[62:]] for x in out.split("\n")[2:]]
+
+    def get_first_unused_and_safely_removable(self):
+        """
+        Returns device subchannel id of the first unused device that
+        does not share all channel path ids with any other used device.
+
+        Requires get_info() to be called first.
+        """
+        used = [x for x in self.devices if x[self.header["Use"]] == "yes"]
+        unused = [x for x in self.devices if x not in used]
+        index = self.header["CHPIDs"]
+        for device in unused:
+            full_chpid_match = [x for x in used
+                                if x[index] == device[index]]
+            if full_chpid_match:
+                continue
+            return device
+        return None
+
+
+class ChannelPaths(object):
+    """
+    Represents channel path info
+    """
+
+    @staticmethod
+    def _split(chpids):
```

Comment:
Could this method level up to independent method in utils_zchannels.py  since it is not relevant to class ChannelPaths?

Reply:
@chunfuwen No, I prefer not to. According to the class design, ChannelPaths should be used for any channel path handling. By pushing it to top, we'd break encapsulation I believe. Quoting [1]:

"To me, most fundamentally, encapsulation means hiding implementation details. Instead of expressing how to do something, we express what to do. I like to say that encapsulation is hiding the “how” with the “what.” Encapsulation is an important concept and an important code quality because it gives us the ability to change the implementation details without affecting other parts of the system."

[1] https://dzone.com/articles/quality-code-is-encapsulated